src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
   simp only [← getLsbD_eq_getElem, getLsbD_replicate]
   by_cases h' : w = 0 <;> simp [h'] <;> omega
 
+@[simp]
+theorem replicate_append_replicate_eq {w n : Nat} {x : BitVec w} (h : w * (n + m) = w * n + w * m := by omega) :
+    x.replicate n ++ x.replicate m = (x.replicate (n + m)).cast h := by
+  apply BitVec.eq_of_getLsbD_eq
+  simp only [getLsbD_cast, getLsbD_replicate, getLsbD_append, getLsbD_replicate]
+  intros i
+  by_cases h₀ : i < w * m <;> by_cases h₁ : i < w * (n + m)
+  · simp only [h₀, decide_true, Bool.true_and, cond_true, h₁]
+  · rw [Nat.mul_add] at h₁
+    simp only [h₀, decide_true, Bool.true_and, cond_true, Bool.iff_and_self, decide_eq_true_eq]
+    omega
+  · have h₂ : i ≥ w * m := by omega
+    simp only [h₀, decide_false, Bool.false_and, show i - w * m < w * n by omega, decide_true,
+      Bool.true_and, cond_false, h₁]
+    congr 1
+    rw [Nat.sub_mul_mod (by omega)]
+  · simp only [h₀, decide_false, Bool.false_and, cond_false, h₁, Bool.and_eq_false_imp,
+    decide_eq_true_eq]
+    omega
+
+@[simp]
+private theorem mod_sub_eq_sub_mod {w n i : Nat} (hwn : i < w * n) (hn : 0 < n) :
```

Comment:
A private simp theorem seems like a bad idea; can you remove this as a simp-lemma and use it explicitly where you need it?

Alternatively, put this as a Nat theorem where it belongs, so it doesn't have to be private

Reply:
I've made it explicit for now, I am afraid it makes the proof much more confused, but let's see

Reply:
How so? It should be as simple as simply adding `mod_sub_eq_sub_mod` to the `simp` invocation that was applying this one implicitly before?

That is, if you first squeeze all your simps (assuming this lemma is not used in a non-squeezed terminal simp), nothing should break if you then remove the simp annotation here

Reply:
oh I see, i misinterpreted the explicitly and deleted the theorem, using its proof directly in `getMsbD_replicate` when necessary

Reply:
but I'm now also refactoring the proof with `cases`, so I'll commit and update once that's done! thanks :)